sentry-core/src/metrics.rs

```diff
@@ -0,0 +1,387 @@
+                MetricType::Set => BucketValue::set_from_str(value),
+                MetricType::Gauge => BucketValue::gauge(value.parse().ok()?),
+            })
+        }
+
+        fn parse(metric: &str) -> Option<(BucketKey, BucketValue)> {
+            let mut components = metric.split('|');
+            let mut values = components.next()?.split(':');
+            let name = values.next()?;
+
+            let ty: MetricType = components.next().and_then(|s| s.parse().ok())?;
+            let mut value = mk_value(ty, values.next()?)?;
+
+            for value_s in values {
+                value.merge(mk_value(ty, value_s)?).ok()?;
+            }
+
+            let mut tags = "";
+            let mut timestamp = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap_or_default()
+                .as_secs();
+
+            for component in components {
+                if let Some(component_tags) = component.strip_prefix('#') {
+                    tags = component_tags;
+                } else if let Some(component_timestamp) = component.strip_prefix('T') {
+                    timestamp = component_timestamp.parse().ok()?;
+                }
+            }
+
+            Some((
+                BucketKey {
+                    timestamp,
+                    ty,
+                    name: name.into(),
+                    tags: tags.into(),
+                },
+                value,
+            ))
+        }
+
+        if let Some(parsed_metric) = parse(metric) {
+            let _ = self.sender.send(Task::RecordMetric(parsed_metric));
+        }
+    }
+
+    pub fn flush(&self) {
+        let _ = self.sender.send(Task::Flush);
+    }
+
+    fn worker_thread(receiver: Receiver<Task>, transport: TransportArc) {
+        let mut buckets = AggregateMetrics::new();
+        let mut last_flush = Instant::now();
+
+        loop {
+            let timeout = FLUSH_INTERVAL.saturating_sub(last_flush.elapsed());
+
+            match receiver.recv_timeout(timeout) {
+                Err(RecvTimeoutError::Timeout) | Ok(Task::Flush) => {
+                    // flush
+                    Self::flush_buckets(std::mem::take(&mut buckets), &transport);
+                    last_flush = Instant::now();
+                }
+                Ok(Task::RecordMetric((mut key, value))) => {
+                    // aggregate
+                    let rounding_interval = FLUSH_INTERVAL.as_secs();
+                    let rounded_timestamp = (key.timestamp / rounding_interval) * rounding_interval;
+
+                    key.timestamp = rounded_timestamp;
+
+                    match buckets.entry(key) {
+                        Entry::Occupied(mut entry) => {
+                            let _ = entry.get_mut().merge(value);
+                        }
+                        Entry::Vacant(entry) => {
+                            entry.insert(value);
```

Comment:
When inserting metrics with high volume, the total size of cached buckets could grow larger than what we can send to the server. In the Python SDK, we therefore count the number of data points inserted with this ruleset:

- gauge creation: +5, all others: +1
- add to set or distribution: +1, others +0

Then, when we cross a hard-coded threshold, we flush immediately after adding to ensure that outgoing envelopes to not grow indefinitely.